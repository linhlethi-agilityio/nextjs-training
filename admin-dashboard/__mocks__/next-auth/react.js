export const useSession = () => ({
  data: {
    user: {
      token: 'access-token',
      email: 'user@gmail.com',
      id: '1',
      name: 'User',
      role: 'admin',
    },
    expires: Date.now().toString(),
  },
  status: 'authenticated',
  update: async () => null,
});
