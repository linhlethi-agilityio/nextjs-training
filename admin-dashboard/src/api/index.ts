export * from './orders';
export * from './customers';
