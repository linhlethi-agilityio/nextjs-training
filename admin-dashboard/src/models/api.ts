export type ResponseData<T> = {
  data: T;
};
