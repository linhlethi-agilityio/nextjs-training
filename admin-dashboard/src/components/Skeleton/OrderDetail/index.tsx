import { Flex, Stack, Heading, Text, Skeleton } from '@chakra-ui/react';

const OrderDetailSkeleton = () => (
  <Flex mt={46} ml={8}>
    <Skeleton width={700} height={300} />
    <Stack flex={1} px={30} pt={4}>
      <Heading fontSize={30}>
        <Skeleton height={10} />
      </Heading>
      <Flex gap={3}>
        <Text>Customer name:</Text>
        <Skeleton height={5} width={100} />
      </Flex>
      <Flex gap={3}>
        <Text>SKU:</Text>
        <Skeleton height={5} width={100} />
      </Flex>
      <Flex alignItems="center" textAlign="center" gap={3} mt={27}>
        <Text>Status:</Text>
        <Skeleton height={4} width={100} borderRadius="2xl" />
      </Flex>
      <Heading mt={27}>
        <Skeleton height={10} width={100} />
      </Heading>
      <Flex gap={3} mt={3}>
        <Text>Deadline:</Text>
        <Skeleton height={4} width={100} />
      </Flex>
      <Skeleton height={4} width={200} mt="auto" />
    </Stack>
  </Flex>
);

export default OrderDetailSkeleton;
