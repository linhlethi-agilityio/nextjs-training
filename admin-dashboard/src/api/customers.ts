import { API_ENDPOINT, DEFAULT_LIMIT } from '@/constants';
import { Customer } from '@/models';
import { api } from '@/services';

interface params {
  limit?: number;
  query?: string;
  page?: number;
}

export const getCustomers = async (params?: params) => {
  try {
    const { limit = DEFAULT_LIMIT, query = '', page = 1 } = params || {};

    const data = await api.getData<Customer[]>(
      API_ENDPOINT.CUSTOMERS,
      {
        limit,
        name: query,
        page,
      },
      {
        next: { tags: ['customers'], revalidate: 3600 },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};

export const getTotalCustomers = async () => {
  try {
    const data = await api.getData<Customer[]>(
      API_ENDPOINT.CUSTOMERS,
      undefined,
      {
        next: { tags: ['customers'], revalidate: 3600 },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};
