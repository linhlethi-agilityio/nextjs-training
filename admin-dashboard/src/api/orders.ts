// Services
import { API_ENDPOINT, DEFAULT_LIMIT, SORT_BY, SORT_ORDER } from '@/constants';

// Models
import { Order } from '@/models';

// Services
import { api } from '@/services';

interface params {
  limit?: number;
  query?: string;
  page?: number;
  sortBy?: SORT_BY;
  sortOrder?: SORT_ORDER;
}

export const getOrders = async (params?: params) => {
  try {
    const {
      limit = DEFAULT_LIMIT,
      query = '',
      page = 1,
      sortBy,
      sortOrder,
    } = params || {};

    const data = await api.getData<Order[]>(
      API_ENDPOINT.ORDERS,
      {
        limit,
        idOrder: query,
        page,
        ...(sortBy && sortOrder && { sortBy, order: sortOrder }),
      },
      {
        next: { tags: ['orders'] },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};

export const getTotalOrders = async () => {
  try {
    const data = await api.getData<Order[]>(API_ENDPOINT.ORDERS, undefined, {
      next: { tags: ['orders'] },
    });

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};

export const getOrderById = async (id: string) => {
  try {
    const data = await api.getData<Order>(
      `${API_ENDPOINT.ORDERS}/${id}`,
      undefined,
      {
        next: { tags: ['order'] },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};
