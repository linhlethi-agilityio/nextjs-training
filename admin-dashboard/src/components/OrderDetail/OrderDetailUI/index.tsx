'use client';

import { useTransition } from 'react';
import dynamic from 'next/dynamic';
import Image from 'next/image';
import {
  Button,
  Flex,
  Heading,
  Stack,
  Text,
  useDisclosure,
} from '@chakra-ui/react';

// Utils
import { formatDateString, getColorByValue } from '@/utils';

// Models
import { Order } from '@/models';

const DynamicOrderModal = dynamic(() => import('../../Modal/OrderModal'));

interface OrderDetailUIProps {
  product: string;
  customer: string;
  status: string;
  deadline: string;
  productImage: string;
  price: number;
  createdAt: string;
  idOrder: string;
  id: string;
  editOrderAction: (
    id: string,
    updateOrder: Partial<Order>,
  ) => Promise<void | string>;
}

const OrderDetailUI = ({
  product,
  customer,
  status,
  deadline,
  id,
  price,
  createdAt,
  idOrder,
  productImage,
  editOrderAction,
}: OrderDetailUIProps) => {
  const { color, bgColor } = getColorByValue(status);
  const [isPending, startTransition] = useTransition();

  const previewData = {
    product,
    customer,
    status,
    deadline,
    price,
    id,
    idOrder,
    createdAt,
    productImage,
  };

  const {
    isOpen: isOpenOrderModal,
    onOpen: onOpenOrderModal,
    onClose: onCloseOrderModal,
  } = useDisclosure();

  const handleEditOrder = (formData: Partial<Order>) => {
    startTransition(async () => {
      const response = await editOrderAction(previewData?.id ?? '', formData);

      if (typeof response !== 'string') {
        onCloseOrderModal();
      }
    });
  };

  return (
    <>
      <Flex display="flex" mt={46} ml={8}>
        <Image src={productImage} alt={product} width={700} height={300} />
        <Stack flex={1} px={30} pt={4}>
          <Heading fontSize={30}>{product}</Heading>
          <Flex gap={3}>
            <Text>Customer name:</Text>
            <Text>{customer}</Text>
          </Flex>
          <Flex gap={3}>
            <Text>SKU:</Text>
            <Text>{idOrder}</Text>
          </Flex>
          <Flex alignItems="center" textAlign="center" gap={3} mt={27}>
            <Text>Status:</Text>
            <Text
              px={15}
              py="5px"
              color={color}
              bgColor={bgColor}
              borderRadius="2xl"
              textAlign="center"
            >
              {status}
            </Text>
          </Flex>
          <Heading mt={27}>${price}</Heading>
          <Flex gap={3} mt={3}>
            <Text>Deadline:</Text>
            <Text>{formatDateString(deadline)}</Text>
          </Flex>
          <Button
            mt="auto"
            maxW={200}
            colorScheme="brand"
            onClick={onOpenOrderModal}
          >
            Edit
          </Button>
        </Stack>
      </Flex>
      {isOpenOrderModal && (
        <DynamicOrderModal
          isLoading={isPending}
          title="Update Order"
          previewData={previewData}
          onSubmitForm={handleEditOrder}
          isOpen={isOpenOrderModal}
          onClose={onCloseOrderModal}
        />
      )}
    </>
  );
};

export default OrderDetailUI;
