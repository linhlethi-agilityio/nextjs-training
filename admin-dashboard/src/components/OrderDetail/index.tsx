import { memo } from 'react';

// Api
import { getOrderById } from '@/api';

// Models
import { Order } from '@/models';

// Component
import OrderDetailUI from './OrderDetailUI';

interface OrderDetailProps {
  id: string;
  editOrderAction: (
    id: string,
    updateOrder: Partial<Order>,
  ) => Promise<void | string>;
}

const OrderDetail = async ({ id, editOrderAction }: OrderDetailProps) => {
  const { data: order } = await getOrderById(id);

  return (
    <OrderDetailUI
      product={order?.product ?? ''}
      customer={order?.customer ?? ''}
      status={order?.status ?? ''}
      deadline={order?.deadline ?? ''}
      price={order?.price ?? 0}
      productImage={order?.productImage ?? ''}
      idOrder={order?.idOrder ?? ''}
      id={order?.id ?? ''}
      createdAt={order?.createdAt ?? ''}
      editOrderAction={editOrderAction}
    />
  );
};

export default memo(OrderDetail);
