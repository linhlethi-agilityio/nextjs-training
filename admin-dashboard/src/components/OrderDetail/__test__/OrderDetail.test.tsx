import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

// Api
import * as api from '@/api';

// Components
import OrderDetail from '..';

jest.mock('@/api');
const mockReplace = jest.fn();
const mockURLSearchParams = {
  page: '1',
};

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  usePathname: jest.fn(),
  useRouter: jest.fn(() => ({ replace: mockReplace })),
  useSearchParams: jest.fn(() => new URLSearchParams(mockURLSearchParams)),
}));

const mockProps = {
  id: '1',
  editOrderAction: jest.fn(),
};

describe('OrderDetail component', () => {
  it('should match snapshot for OrderDetail', async () => {
    jest.spyOn(api, 'getOrderById').mockResolvedValue({
      data: {
        createdAt: '2024-07-23T00:33:27.809Z',
        idOrder: '969-875',
        product: 'fdgf',
        customer: 'Angie McCullough',
        status: 'Rejected',
        deadline: '2024-07-31',
        price: 3244,
        id: '1',
        customerId: '11',
        productImage:
          'https://images.pexels.com/photos/4041392/pexels-photo-4041392.jpeg?cs=srgb&dl=pexels-karolina-grabowska-4041392.jpg&fm=jpg',
      },
    });

    const OrderDetailRender = await OrderDetail({
      ...mockProps,
    });

    const { container } = render(OrderDetailRender);

    expect(container).toMatchSnapshot();
  });
});
