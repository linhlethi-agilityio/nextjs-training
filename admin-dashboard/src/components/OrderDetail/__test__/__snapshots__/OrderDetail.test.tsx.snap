// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetail component should match snapshot for OrderDetail 1`] = `
<div>
  <div
    class="css-1xcuh3p"
  >
    <img
      alt="fdgf"
      data-nimg="1"
      decoding="async"
      height="300"
      loading="lazy"
      src="http://localhost/_next/image?url=https%3A%2F%2Fimages.pexels.com%2Fphotos%2F4041392%2Fpexels-photo-4041392.jpeg%3Fcs%3Dsrgb%26dl%3Dpexels-karolina-grabowska-4041392.jpg%26fm%3Djpg&w=1920&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fimages.pexels.com%2Fphotos%2F4041392%2Fpexels-photo-4041392.jpeg%3Fcs%3Dsrgb%26dl%3Dpexels-karolina-grabowska-4041392.jpg%26fm%3Djpg&w=750&q=75 1x, /_next/image?url=https%3A%2F%2Fimages.pexels.com%2Fphotos%2F4041392%2Fpexels-photo-4041392.jpeg%3Fcs%3Dsrgb%26dl%3Dpexels-karolina-grabowska-4041392.jpg%26fm%3Djpg&w=1920&q=75 2x"
      style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
      width="700"
    />
    <div
      class="chakra-stack css-xzhgh0"
    >
      <h2
        class="chakra-heading css-qhp3qs"
      >
        fdgf
      </h2>
      <div
        class="css-blvggq"
      >
        <p
          class="chakra-text css-0"
        >
          Customer name:
        </p>
        <p
          class="chakra-text css-0"
        >
          Angie McCullough
        </p>
      </div>
      <div
        class="css-blvggq"
      >
        <p
          class="chakra-text css-0"
        >
          SKU:
        </p>
        <p
          class="chakra-text css-0"
        >
          969-875
        </p>
      </div>
      <div
        class="css-177mg6s"
      >
        <p
          class="chakra-text css-0"
        >
          Status:
        </p>
        <p
          class="chakra-text css-gegbys"
        >
          Rejected
        </p>
      </div>
      <h2
        class="chakra-heading css-1178ru4"
      >
        $
        3244
      </h2>
      <div
        class="css-ffrhzl"
      >
        <p
          class="chakra-text css-0"
        >
          Deadline:
        </p>
        <p
          class="chakra-text css-0"
        >
          July 31, 2024
        </p>
      </div>
      <button
        class="chakra-button css-1k591dk"
        type="button"
      >
        Edit
      </button>
    </div>
  </div>
</div>
`;
